Feature References audit — Authentication (#973)
meilisearch/documentation: reference/features/authentication.md

# Authentication

MeiliSearch uses key-based authentication.

When using MeiliSearch in `production`, authentication is mandatory. You can read more about `production` mode in the [configuration reference](/reference/features/configuration).

If a MeiliSearch instance does not use authentication, all routes will be publicly accessible and unprotected.

## Key types

MeiliSearch uses three types of keys:

- The **Master** key grants access to all routes
- The **Private** key grants access to all routes except the `/keys` routes
- The **Public** key only grants access to the following routes:
  - `GET /indexes/:index_uid/search`
  - `POST /indexes/:index_uid/search`
  - `GET /indexes/:index_uid/documents`
  - `GET /indexes/:index_uid/documents/:doc_id`

When a master key is provided, both the private and the public keys are automatically generated. **You cannot create any additional keys**.

> The only unprotected route available when using MeiliSearch in `production` is `GET /health`.

Reviewer: ```suggestion
The only unprotected route available when using the built-in key-based authorization is `GET /health`.
```
Do we want to specify `production` since it can happen in `development` as soon as a master key has been configured ?

> The only unprotected route available when using MeiliSearch in `production` is `GET /health`.

Author: Regarding consistency, shouldn't we also add links to the routes accessible with a public key (lines #16-19)? Or would that be a bit too much?

> The only unprotected route available when using MeiliSearch in `production` is `GET /health`.

Reviewer: It all depends on your guideline about the presence or not of links pointing to another documentation pages. Personally, I like to have links to referenced pages within a documentation to move faster around a subject. 🙂

> ## Adding a key

Reviewer 2: ```suggestion
## Adding the master key
```

When launching a MeiliSearch instance, you have the option of giving it a master key. By doing so, all routes will be protected and will require a key to be accessed.

You can specify a key by passing the `MEILI_MASTER_KEY` environment variable, or using the command line argument `--master-key`.

After adding the master key, you can retrieve both the private and the public keys on the [keys route](/reference/api/keys.md).

> ## Using an authenticated MeiliSearch instance

Reviewer: ```suggestion
## Communicate with a protected instance
```
I suggest something like this because the Meilisearch instance is not actually authenticated, it is the client that communicates with it that will authenticate itself through the header authentication key when it makes a request.

Reviewer: I would say yes!

> You can specify a key by passing the `MEILI_MASTER_KEY` environment variable, or using the command line argument `--master-key`.

Reviewer 2: ````suggestion
You can specify the master key by setting the `MEILI_MASTER_KEY` environment variable, or passing the command line argument `--master-key` on launch.

#### Example

:::: tabs

::: tab Env

```bash
export MEILI_MASTER_KEY=whatever_you_want
meilisearch
```

:::

::: tab CLI

```bash
meilisearch --master-key=whatever_you_want
```

:::

::::
````
Not 100% sure I typed that right, but should create an example code block with two tabs, one for env and one for CLI. If you go with this suggestion, verify that it displays correctly before merging :)

Reviewer 2: Could also ignore this for now and add it in a separate PR if preferred.

Author: sounds great, but one minor comment: what do you say we replace `whatever_you_want` with something like `[YOUR_MASTER_KEY]`? Sure, people could use whatever string they want, but generally we'd like them to use something that's hard to guess and brute-force, no?

Reviewer 2: Yes, great suggestion :)

> After adding the master key, you can retrieve both the private and the public keys on the [keys route](/reference/api/keys.md).

Reviewer 2: ```suggestion
After setting up the master key, you can retrieve both the private and the public keys with the [keys route](/reference/api/keys.md).
```

When using authentication, a key must be added to [the header](/reference/api/README.md#authentication) of each API call.

> We strongly recommend only using the master key to retrieve the public and private keys.

Reviewer 2: ```suggestion
We strongly discourage using the master key for API calls. It is intended only for retrieving the public and private keys.
```

If an invalid key is provided, you will receive the `HTTP/1.1 403 Forbidden` status code. You will receive the same error if you fail to provide a key when querying a protected route.

## Changing a key

**Changing the master key will automatically generate new private and public keys**. It is not possible to change one key without altering the others.

> After changing the master key, it is mandatory to restart the MeiliSearch instance to ensure the renewal of the private and the public keys.

Reviewer 2: ```suggestion
After changing the master key, it is mandatory to restart the MeiliSearch instance to generate new private and public keys.
```
